File: core/java/src/main/java/hara/lang/protocol/IEncodable.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(namespace = "std.protocol.iencodable", name = "IEncodable")
public interface IEncodable {
  @HaraMethod(value = "encode-with", arity = 2)
  Object encodeWith(Object visitor);
}

File: core/rust/src/lang/protocol/iencodable.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.iencodable", name = "IEncodable")]
pub trait IEncodable<V> {
    type Output;

    #[hara_method(value = "encode-with", arity = 2)]
    fn encode_with(&self, visitor: V) -> Self::Output;
}
